src/sentry/api/serializers/rest_framework/dashboard.py

```diff
@@ -0,0 +1,186 @@
+    conditions = serializers.CharField(required=False)
+    interval = serializers.CharField(required=False)
+
+    required_for_create = {"fields", "conditions"}
+
+    validate_id = validate_id
+
+    def validate(self, data):
+        if not data.get("id"):
+            keys = set(data.keys())
+            if keys.intersection(self.required_for_create) != self.required_for_create:
+                raise serializers.ValidationError("The fields and conditions fields are required")
+        return data
+
+
+class DashboardWidgetSerializer(serializers.Serializer):
+    # Is a string because output serializers also make it a string.
+    id = serializers.CharField(required=False)
+    title = serializers.CharField(required=False)
+    displayType = serializers.ChoiceField(
+        choices=DashboardWidgetDisplayTypes.as_text_choices(), required=False
+    )
+    queries = DashboardWidgetQuerySerializer(many=True, required=False)
+
+    def validate_displayType(self, display_type):
+        return DashboardWidgetDisplayTypes.get_id_for_type_name(display_type)
+
+    validate_id = validate_id
+
+    def validate(self, data):
+        if not data.get("id") and not data.get("queries"):
+            raise serializers.ValidationError("One or more queries are required to create a widget")
+        return data
+
+
+class DashboardDetailsSerializer(serializers.Serializer):
+    # Is a string because output serializers also make it a string.
+    id = serializers.CharField(required=False)
+    title = serializers.CharField(required=False)
+    widgets = DashboardWidgetSerializer(many=True, required=False)
+
+    validate_id = validate_id
+
+    def update(self, instance, validated_data):
+        """
+        Update a dashboard, the connected widgets and queries
+
+        - Widgets in the dashboard currently, but not in widget_data will be removed.
+        - Widgets without ids will be created.
+        - Widgets with matching IDs will be updated.
+        - The order of the widgets will be updated based on the order in the request data.
+
+        Only call save() on this serializer from within a transaction or
+        bad things will happen
+        """
+        instance.title = validated_data.get("title", instance.title)
+        instance.save()
+
+        if "widgets" in validated_data:
+            self.update_widgets(instance, validated_data["widgets"])
+
+        return instance
+
+    def update_widgets(self, instance, widget_data):
+        widget_ids = [widget["id"] for widget in widget_data if "id" in widget]
+
+        existing_widgets = DashboardWidget.objects.filter(dashboard=instance, id__in=widget_ids)
+        existing_map = {widget.id: widget for widget in existing_widgets}
+
+        # Remove widgets that are not in the current request.
+        self.remove_missing_widgets(instance.id, widget_ids)
+
+        # Get new ordering start point.
+        next_order = get_next_dashboard_order(instance.id)
```

Comment:
Can't this just always start at 0 since this is a `PUT`? Plus all widgets will be updated so the order for every widget will increase on every save, so this runs the risk of overflow (though it'll take quite a while).

Reply:
I'm going to try that. This code was here from the last implementation, it might not be needed anymore as all the widgets will be reordered.

Reply:
Because of the unique constraints on `dashboard_id, order` we can't reuse values. Postgres checks constraints inside of a transaction as well. We could remove the constraint. That would make this code simpler but would allow potentially inconsistent result orders to occur.

Reply:
If bulk updates were possible, it'd be possible to do this by

1. delete all missing widgets at once
2. update all changed widgets at once
3. insert all new widgets at once

Though it looks like `bulk_update` isn't available in this version of Django. And it's the updates where this constraint is problematic.